util/PromiseQueue.js

```diff
@@ -2,9 +2,18 @@ const PromiseQueue = () => {
   let promise = Promise.resolve()
 
   return {
-    push: (fn) => {
+    push: fn => {
       promise = promise.then(fn, fn)
-      return this
+    },
+    fetch: fn => {
+      let returnResolver
+      const returnPromise = new Promise(resolve => {
+        returnResolver = resolve
+      })
+      promise = promise.then(fn, fn).then(result => {
```

Comment:
Nice work on this.

Just one thing. If `fn` throws then `returnPromise` will never resolve. Just add this to fix it:

```js
promise = promise.then(fn, fn).then(res => returnResolver(res), err => returnRejecter(err))
```